server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +412,300 @@ public int hashCode() {
+            threadpool.schedule(cacheCleaner, cleanInterval, ThreadPool.Names.SAME);
+        }
+
+        /**
+         * Enqueue cleanup key.
+         *
+         * @param cleanupKey the cleanup key
+         */
+        void enqueueCleanupKey(CleanupKey cleanupKey) {
+            keysToClean.add(cleanupKey);
+            updateStaleKeysCount(cleanupKey);
+        }
+
+        /**
+         * Updates the cleanupKeyToCountMap with the given CleanupKey.
+         * If the ShardId associated with the CleanupKey does not exist in the map, a new entry is created.
+         * The method increments the count of the CleanupKey in the map.
+         * <p>
+         * Why use ShardID as the key ?
+         * CacheEntity mainly contains IndexShard, both of these classes do not override equals() and hashCode() methods.
+         * ShardID class properly overrides equals() and hashCode() methods.
+         * Therefore, to avoid modifying CacheEntity and IndexShard classes to override these methods, we use ShardID as the key.
+         *
+         * @param cleanupKey the CleanupKey to be updated in the map
+         */
+        private void updateCleanupKeyToCountMapOnCacheInsertion(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {} while cleaning Indices Request Cache", cleanupKey.readerCacheKeyId);
+                return;
+            }
+            ShardId shardId = indexShard.shardId();
+
+            // If the key doesn't exist, it's added with a value of 1.
+            // If the key exists, its value is incremented by 1.
+            cleanupKeyToCountMap.computeIfAbsent(shardId, k -> ConcurrentCollections.newConcurrentMap())
+                .merge(cleanupKey.readerCacheKeyId, 1, Integer::sum);
+        }
+
+        private synchronized void updateCleanupKeyToCountMapOnCacheEviction(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {} while cleaning Indices Request Cache", cleanupKey.readerCacheKeyId);
+                return;
             }
+            ShardId shardId = indexShard.shardId();
+            // If the key doesn't exist, ignore
+            ConcurrentMap<String, Integer> keyCountMap = cleanupKeyToCountMap.get(shardId);
+            if (keyCountMap != null) {
+                keyCountMap.compute(cleanupKey.readerCacheKeyId, (key, currentValue) -> {
+                    // If the key is not present, no action is needed, return null.
+                    if (currentValue == null) return null;
```

Comment:
You mean
`if (key ==null)` here? I am guessing if currentValue is null, it basically implies key is also null?

Better to refactor/simplify to below?
```
// Only do computation if key exists? Otherwise skip.
keyCountMap.computeIfPresent(cleanupKey.readerCacheKeyId, (key, currentValue) -> {
    staleKeysCount.decrementAndGet(); 
    int newValue = currentValue - 1; 
    return newValue == 0 ? null : newValue;
});
```

Reply:
Yes, you have the refactored version.